### ISSUE TITLE
TensorFlow Custom loop gives different value of loss

### ISSUE THREAD

**Issue by Debottam** [tensorflow/tensorflow#43312]
I am using following versions:

```
tf version:  2.3.0
keras version:  2.4.0
```
my customized error is

```
def mse_fn(y_true, y_pred):
  error = y_true - y_pred
  mserror = tf.reduce_mean(tf.square(error))
  return mserror
```
data processing for my code is

```
(X_train, Y_train), (X_test, Y_test) = keras.datasets.mnist.load_data()
Y_train = np.array(Y_train).astype(np.float32)
Y_test = np.array(Y_test).astype(np.float32)
X_train = X_train/255.
X_test = X_test/255.
```
my customized DNN layer is

```
class MyDense(keras.layers.Layer):
  def __init__(self, units, activation=None, **kwargs):
    super().__init__(**kwargs)
    self.units = units
    self.activation = keras.activations.get(activation)
  
  def build(self, batch_input_shape):
    self.kernel = self.add_weight(name = "kernel", shape=[batch_input_shape[-1], self.units], initializer="glorot_normal")
    self.bias = self.add_weight(name = "bias", shape=[self.units], initializer = 'zeros')
    super().build(batch_input_shape)

  def call(self, X):
    return self.activation(X @ self.kernel + self.bias)
  
  def compute_output_shape(self, batch_input_shape):
    return tf.TensorShape(batch_input_shape.as_list()[:-1]+[self.units])
  
  def get_config(self):
    base_config = super().get_config()
    return {**base_config, "units":self.units, "activation": keras.activations.serialize(self.activation)}

```
the model is

```
class DNN_model(keras.Model):
  def __init__(self, output_dim, **kwargs):
    super().__init__(**kwargs)
    self.hidden1 = keras.layers.Flatten(input_shape = [28, 28])
    self.hidden2 = MyDense(32, activation='relu')
    self.hidden3 = MyDense(16, activation='relu')
    self.hidden4 = MyDense(8, activation='relu')
    self.hidden5 = MyDense(output_dim, activation='relu')
  
  def call(self, inputs):
    Z = self.hidden1(inputs)
    Z = self.hidden2(Z)
    Z = self.hidden3(Z)
    Z = self.hidden4(Z)
    Z = self.hidden5(Z)
    return Z

```
when I compile and fit using model.fit I get following loss

```
model_cust_loss = DNN_model(1)
model_cust_loss.compile(optimizer='adam', loss = mse_fn, metrics=['accuracy'])
model_cust_loss.fit(X_train, Y_train, epochs=20 , batch_size=32)
```
the output

```
1875/1875 [==============================] - 3s 1ms/step - loss: 2.5784 - accuracy: 0.1690
Epoch 2/20
1875/1875 [==============================] - 3s 1ms/step - loss: 1.0625 - accuracy: 0.1910
Epoch 3/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.8146 - accuracy: 0.1947
Epoch 4/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.7062 - accuracy: 0.1981
Epoch 5/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.6342 - accuracy: 0.1999
Epoch 6/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.5785 - accuracy: 0.2014
Epoch 7/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.5263 - accuracy: 0.2015
Epoch 8/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.4914 - accuracy: 0.2036
Epoch 9/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.4480 - accuracy: 0.2066
Epoch 10/20
...
```
and then I try customized loop

```
def random_batch(X, y, batch_size=32):
  idx = np.random.randint(len(X), size=batch_size)
  return X[idx], y[idx]

n_epochs = 5
batch_size = 32
n_steps = len(X_train)//batch_size
optimizer = keras.optimizers.Adam() 
print("n_steps: ", n_steps)

epoch_losses = []
for epoch in range (1,n_epochs+1):
  print("Epoch {}/{}".format(epoch, n_epochs))
  batch_losses = []
  for step in range(1, n_steps+1):
    #print("step:",step)
    X_batch, Y_batch = random_batch(X_train, Y_train, batch_size = 32)
    with tf.GradientTape() as tape:
      Y_pred = model_cust_loss(X_batch, training=True)
      current_loss = mse_fn(Y_batch, Y_pred) 
    batch_losses.append(current_loss)
    gradients = tape.gradient(current_loss, model_cust_loss.trainable_variables)
    optimizer.apply_gradients(zip(gradients,model_cust_loss.trainable_variables))
  epoch_losses.append(np.mean(batch_losses))
  print("loss per epoch:", np.mean(batch_losses))
```
The output gives a different value of loss than the model.fit method

```
Epoch 1/5
loss per epoch: 8.3357525
Epoch 2/5
loss per epoch: 8.387993
Epoch 3/5
loss per epoch: 8.356688
Epoch 4/5
loss per epoch: 8.362662
Epoch 5/5
loss per epoch: 8.374953
```
could you help?

Thanks, Debottam

**Comment by bhack** [comment]
Please, can you clean up a little bit your example so that we can quickly copy, paste, run your code.
Thanks

**Comment by Debottam** [comment]
Hi @bhack , @amahendrakar 
I have created an example in github you should be able to copy and run as it is https://github.com/Debottam/tensorFlowCustomized/blob/master/customizedTensorFlow.ipynb

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with TF v2.3 and TF-nightly. Please find the gist of it [here](https://colab.research.google.com/gist/amahendrakar/fb8218a6ea60aaabea14c8e304b600f3/43312-tf-nightly.ipynb). Thanks!

**Comment by Debottam** [comment]
Hi @amahendrakar , 
Could you provide a solution to this issue?
Thanks,
Debottam.

**Comment by Debottam** [comment]
Hi @jvishnuvardhan ,
Could you help me on this?

Thanks,
Debottam

**Comment by Debottam** [comment]
Dear tensorflow developers,
I need help in this matter.
Thanks,
Debottam

**Comment by bhack** [comment]
@Debottam I've not seen your model and data in detail but it seems to me very similar to the working example at https://colab.research.google.com/github/tensorflow/docs/blob/master/site/en/tutorials/customization/custom_training_walkthrough.ipynb
Can you give a  run to that tutorial?

**Comment by Debottam** [comment]
Hi @bhack 
The loss of https://colab.research.google.com/github/keras-team/keras-io/blob/master/guides/ipynb/writing_a_training_loop_from_scratch.ipynb matches with the 

```
model_cust_loss.fit(X_train, Y_train, epochs=20 , batch_size=32)
```
part of my code which is 

```
1875/1875 [==============================] - 3s 1ms/step - loss: 2.5784 - accuracy: 0.1690
Epoch 2/20
1875/1875 [==============================] - 3s 1ms/step - loss: 1.0625 - accuracy: 0.1910
Epoch 3/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.8146 - accuracy: 0.1947
Epoch 4/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.7062 - accuracy: 0.1981
Epoch 5/20
1875/1875 [==============================] - 3s 1ms/step - loss: 0.6342 - accuracy: 0.1999
```
but the loss I get from my customized loop is way bigger. That is my question and would like to know why its happening?

Thanks,
Debottam.

**Comment by bhack** [comment]
It doesn't seems to me a bug. But if you want to have a specific support on your model/data generation I suggest you to post your question on our channel at:
https://stackoverflow.com/questions/tagged/tensorflow.

Tickets are  mainly for Bug reports and Feature requests.

**Comment by Debottam** [comment]
@bhack ,

Did that too https://stackoverflow.com/questions/63941144/tensorflow-custom-loop-gives-different-value-of-loss didn't get any response.

Thanks,
Debottam.

**Comment by jvishnuvardhan** [comment]
@Debottam I quickly checked and looks like there was a shape mismatch between `y_pred` and `y_true` depending on whether `y_pred` is from model.fit or custom training. I added a `reshape` command as shown below and the error are similar from both the approaches.

`Y_pred = tf.reshape(Y_pred,(1,batch_size))`

Please take a look at the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/ea88b6046ba08b12f0d0ecb0314d8ca7/43312-tf-nightly.ipynb). Thanks!

Ouput from the custom training

```
n_steps:  1875
Epoch 1/20
loss per epoch: 2.3010604
Epoch 2/20
loss per epoch: 1.0047752
Epoch 3/20
loss per epoch: 0.79963183
Epoch 4/20
loss per epoch: 0.68781537
Epoch 5/20
loss per epoch: 0.5867094
Epoch 6/20
loss per epoch: 0.527706
Epoch 7/20
loss per epoch: 0.4982657
Epoch 8/20
loss per epoch: 0.45960277
Epoch 9/20
loss per epoch: 0.43102258
Epoch 10/20
loss per epoch: 0.39782792
Epoch 11/20
loss per epoch: 0.3768253
Epoch 12/20
loss per epoch: 0.37661105
Epoch 13/20
loss per epoch: 0.34664243
Epoch 14/20
loss per epoch: 0.35457188
Epoch 15/20
loss per epoch: 0.3290785
Epoch 16/20
loss per epoch: 0.31267858
Epoch 17/20
loss per epoch: 0.30636093
Epoch 18/20
loss per epoch: 0.29752013
Epoch 19/20
loss per epoch: 0.27903932
Epoch 20/20
loss per epoch: 0.2706806
```

**Comment by Debottam** [comment]
Hi @jvishnuvardhan ,
My apologies for the delayed reply. Thanks a lot for debugging it. 

Thanks,
Debottam.
